### PR TITLE
add the option to crop silence segments in training set curation

### PR DESCRIPTION
**Pull request by yardencsGitHub**
First we try to crop from the end of the dataset
Then, we try from the beginning.
If that doesn't work we crop from silence segments larger than the window_size.

This is done by preventing training windows from starting at points within the silence segments.
Since silence segments can follow non-silence segments or precede them the amount of data being ignored actually depend on whether the silence segment happens to be right in the beginning of the file or right in the end. We make sure to account for that.

### CODE REVIEW

**Comment by NickleDave** [comment]
@yardencsGitHub I think I see at a high level what you're doing here, and I'm willing to merge in if it makes it possible for you to create training sets of smaller duration *and* we're sure it works without introducing bugs in anything else

Next time please let's discuss proposed enhancements, and how we might implement them, *before* you go off and work a whole bunch on your own.

I made some changes for clarity; they should not have changed the behavior of what you added (I hope).

But there's a couple things that are not clear to me. Main things are
- [ ] why specifically window size + 2 time bins? Right now this seems like a ["magic number"](https://en.wikipedia.org/wiki/Magic_number_(programming)) in the code
- [ ] we also have a "magic number" where we assume that there *are* unlabeled segments, and that their label is zero. Instead of this I think we should do the following:
  + check whether 
     ```Python
     'unlabeled' in labelmap
     ````
      which we already pass in to the `crop` method as a parameter
  + if so, put it in a variable and use it in the code block where you find the onsets and offsets of silent periods
  + if not, fail gracefully by raising an informative error

I think it might be easier to discuss face-to-face instead of via messages on a PR. I'll send you an email -- let's talk by Zoom tomorrow sometime

**Comment by NickleDave** [comment]
@yardencsGitHub this last commit adds changes we just went through

please do test as discussed and then once you think it's working, I will squash the fixup commits into yours, then merge in

**Comment by yardencsGitHub** [comment]
Summary of tests with the last commit:

- Ran 'learncurve' with a total training data set of 24000 sec from canary llb16

1.  I was able to create subsets of 150 sec or larger. 
Without this branch, when only allowing cropping from the beginning or end of the dataset, I could reliably subset 360 sec training sets or larger.

2. When requiring datasets too small for reliable curation I was able to 'exit' the runs in known points involving the validity checks for the desired cropping.

**Comment by NickleDave** [comment]
Thank you @yardencsGitHub -- I am going to be extra paranoid and re-run a couple of the BirdsongRepository learning curves to test further. But that should finish up today and then I'll squash + merge in

**Comment by NickleDave** [comment]
I re-ran with `Bird00` and `Bird01` from the `BirdsongRecognition` dataset. Neither crashed.
I did not validate the results from `Bird00` but ran with all dataset durations used previously, `[ 30, 45, 75, 120, 180, 480,]`

For `Bird01` I ran with just `[30, 45, 75]` then inspected the metrics csv after the `learncurve` ran without error. Accuracy was ~95%, roughly the same as before.

There is still some prettying up of the code that could be done but I will resist the urge and just merge in, so we can move forward with other things.
